<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [33]:
import pandas as pd
import psycopg2

In [34]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [35]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [36]:
# текст запроса

query_3_1 = f'''
    SELECT COUNT(name) as "количество вакансий"
    FROM vacancies
'''




In [37]:
df = pd.read_sql_query(query_3_1, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [38]:
query_3_2 = f'''
    SELECT COUNT(name) as "количество работодателей"
    FROM employers
'''

In [39]:
df = pd.read_sql_query(query_3_2, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [40]:
query_3_3 = f'''
    SELECT COUNT(name) as "количество регионов"
    FROM areas
'''

In [41]:
df = pd.read_sql_query(query_3_3, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [42]:
query_3_4 = f'''
    SELECT COUNT(name) as "количество сфер деятельности"
    FROM industries
'''

In [43]:
df = pd.read_sql_query(query_3_4, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


количество сфер деятельности
0                           294

***

*По предварительному анализу можно отразить следующее:*
- Количество регионов **1362**
- Количество работодателей **23501**
- Количество сфер деятельности **294**
- Количество вакансий **49197**

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [44]:
query_4_1 = f'''
    SELECT
        a.name as area,
        COUNT(v.name) as cnt
    FROM vacancies v
    JOIN areas a ON v.area_id = a.id
    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 5
'''

In [45]:
df = pd.read_sql_query(query_4_1, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [46]:
query_4_2 = f'''
    SELECT count(name)
    FROM vacancies
    WHERE salary_from is not NULL OR salary_to is not NULL
    
'''

In [47]:
df = pd.read_sql_query(query_4_2, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [48]:
query_4_3 = f'''
    SELECT 
        round(AVG(salary_from)) as "средние значения для нижней границы зп",
        round(AVG(salary_to)) as "средние значения для верхней границы зп"
    FROM public.vacancies
'''

In [49]:
df = pd.read_sql_query(query_4_3, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


средние значения для нижней грани  средние значения для верхней гран
0                            71065.0                           110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [50]:
query_4_4 = f'''
SELECT
    schedule,
    employment,
    COUNT(*)
FROM public.vacancies
GROUP BY 1, 2
ORDER BY 3 DESC
OFFSET 1 LIMIT 1
'''

In [51]:
df = pd.read_sql_query(query_4_4, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [52]:
query_4_5 = f'''
SELECT
    experience,
    count(*)
FROM public.vacancies
GROUP BY 1
ORDER BY 2
'''

In [53]:
df = pd.read_sql_query(query_4_5, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

**Выводы по детальному анализу вакансий:**
- Самое большое количество вакансий в Москве, далее идет Санкт-Петербург и Минск.
- Всего примерно 50% от всех вакансий указана хотя бы минимальная или максимальная зарплатная оплата, остальные обсуждают во время собеседования.
- От 71065 до 110537 среднее значение предлагаемой заработной платы. 
- Полный день/Полная занятость и Удаленная работа/Полная занятость - это самые востребованые запросы по типу трудоустройства в сочетании с рабочим графиком. 
- Чаще всего в вакансиях указывают стаж работы от 1 года до 3 лет, далее идет от 3 лет до 6 лет. Реже всего запрашивают большие стажи работ. 


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [54]:
query_5_1 = f'''
SELECT
    e.name,
    count(v.name)
FROM public.employers e
JOIN public.vacancies v ON v.employer_id = e.id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''

In [55]:
df = pd.read_sql_query(query_5_1, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [56]:
query_5_2 = f'''
SELECT
    a.name area,
    count(e.id) cnt_e,
    count(v.id) cnt_v
FROM public.areas a
LEFT JOIN public.vacancies v ON v.area_id = a.id
LEFT JOIN public.employers e ON e.area = a.id
where v.id is null
GROUP BY a.id
ORDER BY 2 DESC
LIMIT 5
'''

In [57]:
df = pd.read_sql_query(query_5_2, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area  cnt_e  cnt_v
0              Россия    410      0
1           Казахстан    207      0
2  Московская область     75      0
3  Краснодарский край     19      0
4            Беларусь     18      0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [58]:
query_5_3 = f'''
SELECT
    e.name,
    COUNT(distinct v.area_id)
FROM public.EMPLOYERS e 
JOIN public.vacancies v ON v.employer_id = e.id
GROUP BY e.id
ORDER BY 2 DESC
LIMIT 5
'''


In [59]:
df = pd.read_sql_query(query_5_3, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0                  Яндекс    181
1              Ростелеком    152
2              Спецремонт    116
3  Поляков Денис Иванович     88
4               ООО ЕФИН      71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [60]:
query_5_4 = f'''
SELECT COUNT(e.id) "Количество без сферы деятельности"
from public.employers_industries ei
RIGHT JOIN public.employers e ON e.id = ei.employer_id
WHERE industry_id is null
'''


In [61]:
df = pd.read_sql_query(query_5_4, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество без сферы деятельности
0                               8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [62]:
query_5_5 = f'''
SELECT
    e.name "Работодатель", 
    count(ei.industry_id) "Количество сфер деятельности"
from public.employers e
JOIN public.employers_industries ei ON ei.employer_id = e.id
GROUP BY e.id
HAVING count(ei.industry_id) = 4
ORDER BY 1
OFFSET 2 LIMIT 1
'''

In [63]:
df = pd.read_sql_query(query_5_5, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Работодатель  Количество сфер деятельности
0         2ГИС                             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [64]:
query_5_6 = f'''
SELECT COUNT(ei.employer_id) "Количество"
from public.employers_industries ei
JOIN public.industries i ON i.id = ei.industry_id
GROUP BY i.name
HAVING i.name = 'Разработка программного обеспечения'
'''

In [65]:
df = pd.read_sql_query(query_5_6,connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество
0        3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [66]:
a.name in ('Москва','Санкт-Петербург','Новосибирск','Екатеринбург','Казань','Нижний Новгород','Челябинск','Красноярск','Самара','Уфа','Ростов-на-Дону','Омск','Краснодар','Воронеж','Пермь','Волгоград')

NameError: name 'a' is not defined

In [77]:
query_5_7 = f'''
select
    a.name,
    count(v.id) cnt
FROM public.areas a 
JOIN public.vacancies v ON v.area_id = a.id
JOIN public.employers e ON e.id = v.employer_id
WHERE e.name = 'Яндекс' AND a.name in ('Москва','Санкт-Петербург','Новосибирск','Екатеринбург','Казань','Нижний Новгород','Челябинск','Красноярск','Самара','Уфа','Ростов-на-Дону','Омск','Краснодар','Воронеж','Пермь','Волгоград')
GROUP BY a.id
UNION ALL
SELECT
    'Total',
    COUNT(*)
FROM public.vacancies v
JOIN public.areas a ON a.id = v.area_id
JOIN public.employers e ON v.employer_id = e.id
WHERE e.name = 'Яндекс' AND a.name in ('Москва','Санкт-Петербург','Новосибирск','Екатеринбург','Казань','Нижний Новгород','Челябинск','Красноярск','Самара','Уфа','Ростов-на-Дону','Омск','Краснодар','Воронеж','Пермь','Волгоград')
ORDER BY 2
'''

In [78]:
df = pd.read_sql_query(query_5_7,connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

**выводы по анализу работодателей:**

- Значительно больше всех вакансий предлагает "Яндекс" с показателем 1933, когда последующие в топ-5 у всех показатель с 400-500.
- В России отмечена самое большое количество работодателей, у которых нет вакансий. 
- ТОП-3 работодателя который предлагает вакансии в разных регионов:
    1. Яндекс;
    2.	Ростелеком;
    3.	Спецремонт.
- 8419 работодатели не указывают сферы деятельности.
- 3553 работодателя указывает сферу деятельности "Разработка программного обеспечения".
- У Яндекса во всех городах-миллиониках есть предлагаемые вакансии, диапозон от 21 до 54, из 485 всего. В топ входит: Москва, Санкт-Петербург и Екатеринбург.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [71]:
query_6_1 = f'''
SELECT count(*) "Количество вакансий"
FROM public.vacancies v 
where v.name ilike '%data%'
    OR v.name ilike '%данн%'
'''

In [72]:
df = pd.read_sql_query(query_6_1, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество вакансий
0                 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [75]:
query_6_2 = f'''
SELECT count( *) "Количество вакансий для начинающих"
FROM public.vacancies v 
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name ilike '%ML%' AND v.name not ilike '%html%')
    
    AND
    
    (v.name ilike '%junior%'
    OR v.experience ilike '%Нет опыта%'
    OR v.employment ilike '%Стажировка%')
'''

In [76]:
df = pd.read_sql_query(query_6_2, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество вакансий для начинающи
0                                 51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [79]:
query_6_3 = f'''
SELECT count(*)
FROM public.vacancies v 
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    
    AND
    
    (v.key_skills like '%SQL%'
    OR v.key_skills ilike '%postgres%')
'''

In [80]:
df = pd.read_sql_query(query_6_3, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [81]:
query_6_4 = f'''
SELECT count(*)
FROM public.vacancies v 
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    
    AND
    
    (v.key_skills ilike '%Python%')
'''

In [82]:
df = pd.read_sql_query(query_6_4, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [89]:
query_6_5 = f'''
SELECT ROUND(AVG(LENGTH(v.key_skills) - LENGTH(replace(v.key_skills, CHR(9), '')) + 1),2)
FROM public.vacancies v 
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    
    AND NOT v.key_skills is NULL
'''

In [90]:
df = pd.read_sql_query(query_6_5, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [87]:
query_6_6 = f'''
SELECT 
    v.experience,
    ROUND(AVG(coalesce((v.salary_from + v.salary_to)/2, v.salary_to, v.salary_from))) avg_salary
FROM public.vacancies v 
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    
    AND
    
    (v.salary_to is not null 
    OR v.salary_from is not null)
    
group by 1
'''

In [88]:
df = pd.read_sql_query(query_6_6, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

**Выводы по предметному анализу:**

- Всего вакансий специалистов по data - 1771
- Для начинающих всего предлагается 51
- Среди которых ключевыми навыками указано SQL - 201 / Python - 351
- 6-7 среднее количество ключевых навыков запрашивается в вакансиях.
- Средняя предлагаемая ЗП для новичка 75000 руб, от 1 до 3 лет опыта 140 000 руб и свыше 3 лет стажа предлагается 243 000 руб.

# Общий вывод по проекту

In [106]:
# Количество вакансий с отношением к данным по регионнам
query_7_1 = f'''
SELECT 
    a.name area,
    count(v.id) "Количество вакансий"
FROM public.vacancies v 
JOIN public.areas a ON a.id = v.area_id
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
GROUP BY 1
ORDER BY 2 desc
LIMIT 10
'''
    
# Количество вакансий с отношением к данным по регионнам для начинающих специалистов
query_7_2 = f'''
SELECT 
    a.name area,
    count(v.id) "Количество вакансий для новичков"
FROM public.vacancies v 
JOIN public.areas a ON a.id = v.area_id
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    AND
    (v.name ilike '%junior%'
    OR v.experience ilike '%Нет опыта%'
    OR v.employment ilike '%Стажировка%')
GROUP BY 1
ORDER BY 2 desc
LIMIT 5
'''

# Наименование работодателей в Москве по количеству вакансий с отношением к данным
query_7_3 = f'''SELECT 
    e.name "Работодатель",
    count(v.id) "Количество вакансий"
FROM public.vacancies v 
JOIN public.employers e ON e.id = v.employer_id
join public.areas a ON v.area_id = e.area
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    AND
    a.name = 'Москва'
GROUP BY 1
ORDER BY 2 desc
LIMIT 5
'''

# По регионам, средняя зарплата среди вакансий с отношением к данным
query_7_4 = f'''
SELECT 
    a.name area,
    v.name vacancies,
    ROUND(AVG(coalesce((v.salary_from + v.salary_to)/2, v.salary_to, v.salary_from))) avg_salary
FROM public.vacancies v 
JOIN public.areas a ON a.id = v.area_id
where (v.name ilike '%data scientist%'
    OR v.name ilike '%data science%'
    OR v.name ilike '%исследователь данных%'
    OR v.name ilike '%machine learning%'
    OR v.name ilike '%машинн%обучен%'
    OR v.name like '%ML%' AND v.name not ilike '%html%')
    AND
    (v.salary_to is not null 
    OR v.salary_from is not null)
    
    
group by v.name, a.name
ORDER BY avg_salary DESC
LIMIT 10
'''








In [95]:
# Количество вакансий с отношением к данным по регионнам
df = pd.read_sql_query(query_7_1, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0           Москва    217
1  Санкт-Петербург     64
2      Новосибирск     23
3  Нижний Новгород     20
4           Казань     16
5           Алматы     16
6            Минск      9
7            Томск      8
8     Екатеринбург      7
9          Армения      6

In [107]:
# Количество вакансий с отношением к данным по регионнам для начинающих специалистов
df = pd.read_sql_query(query_7_2, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area  Количество вакансий для новичков
0           Москва                                19
1  Санкт-Петербург                                 9
2      Новосибирск                                 5
3     Екатеринбург                                 2
4            Минск                                 2

In [108]:
# Наименование работодателей в Москве по количеству вакансий с отношением к данным
df = pd.read_sql_query(query_7_3, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Работодатель  Количество вакансий
0            СБЕР                   27
1              VK                    9
2  Банк ВТБ (ПАО)                    6
3             МТС                    6
4        inDriver                    5

In [109]:
# По регионам, средняя зарплата среди вакансий с отношением к данным
df = pd.read_sql_query(query_7_4, connection)
df

c:\Users\sabvip\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area                                          vacancies  \
0           Москва                       ML Engineer \ Data scientist   
1           Москва  ML & Advanced analytics Lead / Lead Data Scien...   
2           Москва                              Senior Data Scientist   
3  Санкт-Петербург                                     Data Scientist   
4  Санкт-Петербург                              Senior Data Scientist   
5             Кипр  Senior Data Analyst / Senior Data Scientist (A...   
6           Москва              Руководитель направления Data Science   
7      Новосибирск                                Lead data scientist   
8          Армения                          Machine Learning Engineer   
9           Москва                                     Data scientist   

   avg_salary  
0    450000.0  
1    350000.0  
2    300000.0  
3    300000.0  
4    300000.0  
5    300000.0  
6    300000.0  
7    300000.0  
8    292242.0  
9    275000.0

**ИТОГ ИССЛЕДОВАНИЯ:**

- Из 49 197 всего вакансий, отношение к данным предлагается 1771; 
- Самое большое количество вакансий в любом вопросе - Москва, из которых "СБЕР" предлагает больше всех вакансий с отношением к данныим;
- Из 51 вакансий для новичков по данным, 19 из них предлагаются в Москве. Больше всех предлагает "The Skolkovo Institute of Science and Tecnology" - 3 штуки; "СтарЛайн, Научно-производственное объединение" - 2 штуки; "БАНК ВТБ" - 2 штуки.
- Самые большие средние уровень зарплаты наблюдаются чаще всего в Москве и для высокостоящих должностей, в зависимости от стажа предлагаемая ЗП увеличивается. Так же можно отметить, что вакансии с отношением к данным предлагается во всех регионах и отмечается высокий оплачиваемый труд;
- Показатель средний уровень заработный платы по вакансиям с отношением к данным выше средней уровень заработной платы среди всех вакансий. Данная сфера деятельности оплачивается чаще выше других сфер. 